### ISSUE TITLE
Scenario Filter: Exclude scenario if title matches string

### ISSUE THREAD

**Issue by darrencocco** [Behat/Behat#1165]
To facilitate automated testing with modular and extendable products. It would be useful to be able to exclude specific scenarios that test a specific piece of functionality that may have been removed.

It would be preferable to do so without having to just blacklist the entire feature file or changing the upstream file and then managing merges and code reviews.

**Comment by jakzal** [comment]
Sounds like a perfect job for tags.

**Comment by darrencocco** [comment]
Tags don't quite do it on a sufficiently granular level without editing the
upstream projects or maintaining a set of patches that specifically tag
every scenario that must excluded.

**Comment by darrencocco** [comment]
I would be happy if the LazyFeatureIterator did not hard code filters and
instead used dependency injection to allow extensions like the rest of the
code base. Then everyone could write their own filters to suit their needs.

**Comment by aheermohsinse** [comment]
At first level, we may use tags to highlight such; later can dump that code pieces at 2nd level.
I will look into the practical example and share a mockup code for it, could we please have a relevant example feature for this, to try hitting solution.

+ are we sure no one would need it later in future?

**Comment by darrencocco** [comment]
@jakzal 
I am building extension B to upstream project A that has behat tests.
There is a feature file in A that has 8 scenarios, 2 of those scenarios are no longer relevant because a part they are testing for has been significantly altered or removed. The other 6 scenarios are just fine and should be retained to ensure the tests are as comprehensive as possible.

Changing the feature file in project A is not feasible and those two specific scenarios don't have tags that clearly identifies just them. Blacklisting the entire feature file is overkill and reduces test coverage because 6 of the scenarios are still valid. As such those scenario will always fail because they cannot be filtered out when extension B is used.

Copying and pasting the files and just keeping the scenarios that work in B and blacklisting the feature file in A will create the risk of a false positive test result in the future as changes may be made to the upstream feature file that will not be reflected in B and test coverage that could be relevant would be lost.

**Comment by darrencocco** [comment]
I am going to write a workaround for this in the mean time(will be used until the functionality exists in the version that the core software that I am building the extension on) using a SpecificationLocator but I consider it more of an ugly workaround rather than an elegant solution. Happy for the team to just take some of the solution I write and use it to implement the feature.

**Comment by spolischook** [comment]
There is no way to skip step on some condition, I have try it to implement but without success.
If you'll have any code to review will happy to do so.

**Comment by darrencocco** [comment]
Going to work on it on Sunday hopefully have something resembling sane code by Wednesday.

**Comment by lchrusciel** [comment]
We have something similar in mind in [FriendsOfBehat](https://github.com/FriendsOfBehat) organization. We want to extend ExcludeSpecificationsExtension to support Behat CLI syntax(https://github.com/FriendsOfBehat/ExcludeSpecificationsExtension/issues/5), but didn't find time yet. For now, you can only exclude whole feature file. @darrencocco maybe you can just help us and provide more generic solution? WDYT?

**Comment by spolischook** [comment]
I'm really interested in skip scenarios by condition, I can help with implementation or review this feature in behat.

**Comment by darrencocco** [comment]
Yeah I found the FriendsOfBehat one already(was trying to avoid re-inventing the wheel). If you take a look at the code in my tree it provides a framework for adding arbitrary filters and re-writes the current way that filters are handled to use that framework(works in suit definitions and CLI). I have never worked on Symfony or Behat code so it might not be up to community standards yet but hopefully this is a good starting point.

It should actually only take me a few hours to put together that kind of filter, I will see if I can put some time aside in the next 48 hours to write an example in my spare time.
